v2/tools/generator/internal/codegen/pipeline/detect_skipping_properties.go

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/astmodel"
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/codegen/storage"
+)
+
+const DetectSkippingPropertiesStageID = "detectSkippingProperties"
+
+// DetectSkippingProperties detects any properties that skip one or more versions of an object and are reintroduced.
+// As described in issue #1776, such properties are a problem because they might create incompatibilities between
+// different versions of ASO, or between different versions of a given resource.
+//
+// We need to design a long term solution to this problem, but in the meantime we just want to detect the problem to
+// ensure we don't inadvertently release an update to ASO with this significant hidden flaw.
+//
+// Detection works by scanning for properties that are dropped between versions of a resource. We keep track of all
+// these properties and can raise an error if a specific property appears more than once (implying there are two or more
+// sequence versions of that property).
+//
+// To illustrate, assume we have the following set of objects across versions of our mythical CRM service:
+//
+// v1: Person(GivenName, FamilyName, Address)
+// v2: Person(KnownAs, FullName, FamilyName)
+// v3: Person(KnownAs, FullName, FamilyName, Address)
+//
+// Scanning these types, we find:
+//
+// FamilyName: present in (v1, v2, v3); no issue
+// FullName: present in (v2, v3); no issue
+// KnownAs: also present in (v2, v3); no issue
+// GivenName: present in (v1); no issue
+//
+// Address: present in (v1, v2); issue to report, aborting the pipeline.
+//
+// Additional complexities:
+// - We need to handle type renaming between versions.
+// - When introduced, we will also need to handle property renaming between versions
+//
+func DetectSkippingProperties() Stage {
+	return MakeStage(
+		DetectSkippingPropertiesStageID,
+		"Detect properties that skip resource or object versions",
+		func(ctx context.Context, state *State) (*State, error) {
+			detector := NewSkippingPropertyDetector(state.Types(), state.ConversionGraph())
+
+			// Add resources and objects to the graph
+			for _, def := range state.Types() {
+				if t, ok := astmodel.AsPropertyContainer(def.Type()); ok {
+					err := detector.AddProperties(def.Name(), t)
+					if err != nil {
+						return nil, err
+					}
+				}
+			}
+
+			err := detector.CheckForSkippedProperties()
+
+			return state, err
+		})
+}
+
+type SkippingPropertyDetector struct {
+	links              map[astmodel.PropertyReference]astmodel.PropertyReference // Individual links in chains of related properties
+	observedProperties *astmodel.PropertyReferenceSet                            // Set of properties we've observedProperties
```

Comment:
minor: // Set of properties we've observed

Reply:
Dammit. Refactoring tooling gone mad! 😁